### ISSUE TITLE
Question about model training

### ISSUE THREAD

**Issue by kuanchiun** [nanoporetech/bonito#175]
Hello!

I would like to build the custom model using bonito because some repeat sequences have basecalling errors in the default DNA model. And I have two questions.

1. I noticed that in bonito training guidelines do not mention how we set validation dataset in the `bonito train`, here is the command which I use to train a custom model.

`bonito basecaller dna_r9.4.1@v3.3 --save-ctc --reference chm13_1.1/chm13.draft_v1.1_breakline.fasta fast5_data > training/ctc_data/basecalls.sam`
`bonito train --epochs 50 --lr 1e-5 --batch 30 --pretrained dna_r9.4.1@v3.3 --directory training/ctc_data training/model_ver1`

and I got an output message like this.
![validation missing](https://user-images.githubusercontent.com/89234204/130174775-1972467b-d8fd-4078-8fb0-aa7f228d2899.PNG)

Is there any parameter in `bonito train` that can let us set our validation dataset?

2. And after generating the custom model, I use another dataset generated in the same way to verify it. I found that basecalling errors in the repeat sequences are corrected, but another problem appears.

Here are the results in which I use porechop to remove adapters in the same sample.
Porechop command  : `porechop -i sample.fastq -o sample_trim.fastq --adapter_threshold 55 --middle_threshold 85 --end_threshold 55 --end_size 50 --extra_end_trim 0 -t 32 -v 3 > sample_porechop.txt`

Guppy with the default model
![porechop in guppy default model](https://user-images.githubusercontent.com/89234204/130175948-a081ea60-0b6e-4c91-94ed-c46734ad33ab.PNG)

Bonito with the default model
![porechop in default model in bonito](https://user-images.githubusercontent.com/89234204/130175984-1820bc7f-502e-4d79-8864-0f8306fb5c8f.PNG)

Bonito with the custom model
![porechop in training model in bonito](https://user-images.githubusercontent.com/89234204/130176003-116bef00-3c7d-4148-935d-bcfb86b1b78c.PNG)

I found that compared with Guppy and bonito's default model, the adapter counts which porechop detect dramatically decrease in the custom model.

My guess is that because the Bonito train is based on the reference genome, and adapter sequences do not exist in the reference genome, which causes the error training result in the custom model.

Is there any method that can help us to avoid this problem? Should I remove the adapter sequence signal in fast5 file?

Thanks,
Ken Tung

**Comment by iiSeymour** [comment]
Hey @89213385 

 1. You can use a specific set of reads as a validation ctc set for training by saving them inside the training directory in a `validation` folder (https://github.com/nanoporetech/bonito/blob/master/bonito/cli/train.py#L41).

```
bonito basecaller dna_r9.4.1@v3.3 --save-ctc --reference ref.fasta f5_train > train/ctc_data/calls.sam
bonito basecaller dna_r9.4.1@v3.3 --save-ctc --reference ref.fasta f5_valid > train/ctc_data/validation/calls.sam
```
 2. Yes, care must be taken around the boundaries of ctc chunks as not to introduce such edge effects. Whilst the `--save-ctc` workflow makes it easy to train and fine tune models it's not the most robust in this regard. You can try turning up the default coverage requirement for a chunk with `--ctc-min-coverage`, the default is 90%, maybe 98/99% would help 
https://github.com/nanoporetech/bonito/blob/master/bonito/cli/basecaller.py#L99. 

```
bonito basecaller dna_r9.4.1@v3.3 --ctc-min-coverage 0.99 --save-ctc --reference ref.fasta f5_train > train/ctc_data/calls.sam
```

HTH

Chris.

**Comment by 89213385** [comment]
Hi @iiSeymour 

Thank you for your reply, I will try to add my validation ctc set in the bonito train command.

But for question 2, actually, I have no experience in model training before. I'm not very sure how does `--ctc-min-coverage` affects model training.
Where should I find information about `--ctc-min-coverage`?

Thanks,
Ken Tung.

**Comment by 89213385** [comment]
Hi @iiSeymour 

Thanks for your reply, I tried to use `--ctc-min-coverage` and got a better custom model.

In addition, I have another question.
Actually, I tried to estimate the length of certain repeat sequences.
When I use the default model in both guppy and bonito to do basecalling, I found that those repeat sequences have some errors. e.g. CCCTAA -> TGGCC.

And I tried to use CHM13 as our reference genome and train custom model, but for those repeat sequences in my sample, the length is much longer than the repeat sequences in CHM13, as shown below.
![bonito_ask](https://user-images.githubusercontent.com/89234204/133539162-a0ca2078-e979-4961-8432-7580e3baad55.PNG)

And I think that for bonito, there are some repeat sequence could not be fully mapped on the reference genome.

Will this situation cause errors in the calculation of CTC data?


Thanks, 
Ken Tung

**Comment by davidnewman02** [comment]
Closing as the original issue has been resolved.

- Reads that are unaligned are excluded from training as they have no reference sequence. 
- If incorrect reference sequences are provided (e.g. using a CHM13 reference on reads which have different numbers of repeats) then this will likely result in a poor model as it will attempt to fit to an incorrect set of labels.
